packages/workspace/src/browser/workspace-service.ts

```diff
@@ -102,31 +267,127 @@ export class WorkspaceService implements FrontendApplicationContribution {
+        const workspaceFolder = workspaceRootChange.root;
+        if (!this.opened || !workspaceFolder) {
+            throw new Error('Folder cannot be added as there is no active folder in the current workspace.');
+        }
+
+        const rootToAdd = uri.toString();
+        const valid = await this.toValidRoot(rootToAdd);
+        if (valid) {
+            if (this._root && !this._workspaceRootChange.roots.find(r => r.uri === valid.uri)) {
+                this._workspaceRootChange.roots.push(valid);
+                const configUri = await this.getWorkspaceConfigFileUri();
+                if (configUri) {
+                    if (!await this.fileSystem.exists(configUri.toString())) {
+                        await this.fileSystem.createFile(configUri.toString());
+                    }
+                    this._workspaceConfigFile = await this.fileSystem.setContent((await this.fileSystem.getFileStat(configUri.toString()))!,
+                        JSON.stringify({ roots: this._workspaceRootChange.roots.map(r => r.uri) })
+                    );
+                    this._workspaceRootChange.workspaceConfigFile = this._workspaceConfigFile;
+                }
+                this.startWatch(valid);
+                this.onWorkspaceRootChangeEmitter.fire(this._workspaceRootChange);
+            }
+            return;
+        }
+        throw new Error(`Invalid workspace root URI. Expected an existing directory location. URI: ${rootToAdd}.`);
+    }
+
+    /**
+     * Removes folder(s) from workspace.
+     */
+    async removeFolders(uris: URI[]): Promise<void> {
+        if (!this.opened) {
+            throw new Error('Folder cannot be removed as there is no active folder in the current workspace.');
+        }
+        const validUris: string[] = [];
+        for (const uri of uris.map(u => u.toString())) {
+            const valid = await this.toValidRoot(uri);
+            if (valid) {
+                validUris.push(uri);
+            }
+        }
+
+        if (validUris.length > 0 && this._root) {
+            const messageContainer = document.createElement('div');
+            messageContainer.textContent = 'Remove the following folders from workspace? (note: nothing will be erased from disk)';
+            const list = document.createElement('ul');
+            list.style.listStyleType = 'none';
+            validUris.forEach(u => {
+                const listItem = document.createElement('li');
+                listItem.textContent = uris.find(uri => uri.toString() === u)!.displayName;
+                list.appendChild(listItem);
+            });
+            messageContainer.appendChild(list);
+            const dialog = new ConfirmDialog({
+                title: 'Remove Folder from Workspace',
+                msg: messageContainer
+            });
+            if (await dialog.open()) {
+                this._roots = this._roots.filter(root => validUris.indexOf(root.uri) < 0);
```

Comment:
it would be nice to have a simple flow instead of updating the state here and there, i.e.:
- change a file
- get changes from fs
- update the state (only one method which can update the workspace description)